website/docs/generics.md

```diff
@@ -0,0 +1,1056 @@
+
+Since `Elem` is invariant (has no explicit variance annotation), Sorbet reports
+an error on the `T.let` attempting to widen the type of `int_box` to
+`Box[Numeric]`. Two objects of a given generic class with an invariant type
+member (`Box` in this example) are only subtypes if the types bound to their
+invariant type members are **equivalent**.
+
+Invariant type members, unlike covariant and contravariant type members, maybe
+be used in **both** input and output positions within method signatures. This
+nuance is explained in more detail in the next sections about covariant and
+contravariant type members.
+
+> **Note**: all type members in a Ruby `class` must be invariant. Only type
+> members in a Ruby `module` are allowed to be covariant or contravariant. See
+> the docs for error code [5016](error-reference.md#5016) for more information.
+
+### Covariance (`:out`)
+
+Covariant type members preserve the subtyping relationship. Specifically, if the
+type `Child` is a subtype of the type `Parent`, then the type `M[Child]` is a
+subtype of the type `M[Parent]` if the type member of `M` is covariant. In
+symbols:
+
+```
+Child <: Parent  ==>  M[Child] <: M[Parent]
+```
+
+Note that only a Ruby `module` (not a `class`) may have a covariant type member.
+(See the docs for error code [5016](error-reference.md#5016) for more
+information.) Here's an example of a module that has a covariant type member:
+
+```ruby
+# covariant `Box` interface
+module IBox
+  extend T::Generic
+  # `:out` declares this type member as covariant
+  Elem = type_member(:out)
+end
+
+sig {params(int_box: IBox[Integer]).void}
+def example(int_box)
+  T.let(int_ibox, IBox[Numeric]) # OK
+end
+```
+
+In this case, the `T.let` assertion reports no static errors because
+`Integer <: Numeric`, and therefore `IBox[Integer] <: IBox[Numeric]`.
+
+Covariant type members may only appear in **output** positions (thus the `:out`
+annotation). For more information about what an output position is, see
+[Input and output positions](#input-and-output-positions) below.
+
+In practice, covariant type members are predominantly useful for creating
+interfaces that produce values of the specified type. For example:
+
+```ruby
+module IBox
+  extend T::Sig
+  extend T::Generic
+  abstract!
+
+  # Covariant type member
+  Elem = type_member(:out)
+
+  # Elem can only be used in output position
+  sig {abstract.returns(Elem)}
+  def value; end
+end
+
+class Box
+  extend T::Sig
+  extend T::Generic
+
+  # Implement the `IBox` interface
+  include IBox
+
+  # Redeclare the type member, to be compatible with `IBox`
+  Elem = type_member
+
+  # Within this class, `Elem` is invariant, so it can also be used
+  sig {params(value: Elem).void}
+  def initialize(value:); @value = value; end
```

Comment:
So that the example fits on smaller screens without excessive scrolling, and so that method bodies are de-emphasized relative to method signatures, which are the important part of this example.